core/src/main/java/org/jruby/RubyClass.java

```diff
@@ -1427,24 +1426,29 @@ public synchronized void reify(String classDumpDir, boolean useChildLoader) {
 
         final byte[] classBytes = reifier.reify();
 
-        final ClassDefiningClassLoader parentCL;
+        ClassDefiningClassLoader classLoader; // usually parent's class-loader
         if (parentReified.getClassLoader() instanceof OneShotClassLoader) {
-            parentCL = (OneShotClassLoader) parentReified.getClassLoader();
+            classLoader = (OneShotClassLoader) parentReified.getClassLoader();
+
+            if (classLoader.hasDefinedClass(javaName)) { // class removed after being reified but parent got kept around
+                // while this seems like a leak the parent Java class might be GCable just hasn't been collected, yet
+                classLoader = new OneShotClassLoader((OneShotClassLoader) classLoader);
```

Comment:
this extra wrapping (which under normal condition won't happen) is the fix, 
not super happy about it but I wonder if there are any concerns?

Reply:
alternatively we could append some suffix (counter) to the class-name but having the same name seemed nice.

this also reminds me that using `rubyobj.` package prefix with modularization might not be "ideal".

if we would to consider having some kind of package prefix than we might want to leverage `org.jruby.gen`:
- "`org.jruby.gen.reified_#{ruby_class.id}.SubList"`

Reply:
alternative proposal: https://github.com/jruby/jruby/pull/8147

Reply:
alternative alternative proposal: https://github.com/jruby/jruby/pull/8185